custom_extension_repository over https://

carlopi opened the pull request:
This PR allows to make extension over https protocol work.

Fixes #9818

Demo is like:
`git checkout 0fd6fb9198 && GEN=ninja make && wget https://carlopi.github.io/extension-repository/0fd6fb9198/osx_arm64/httpfs.duckdb_extension`
To build the duckdb binary + download httpfs extension locally (otherwise also building it it's fine).

Then:
```sql
carlo@ScroogeMcDuck duckdb2 % ./build/release/duckdb -unsigned      
v0.9.3-dev897 0fd6fb9198
Enter ".help" for usage hints.
D SET extension_directory = 'empty-extension-folder';
D SET custom_extension_repository='https://carlopi.github.io/extension-repository';
Error: Invalid Input Error: httpfs must be already locally installed before changing to a custom_extension_repo that requires httpfs to be accessed
D INSTALL 'httpfs.duckdb_extension';
D SET custom_extension_repository='https://carlopi.github.io/extension-repository';
D FORCE INSTALL icu;
Error: IO Error: Failed to download extension "icu" at either of:
     "https://carlopi.github.io/extension-repository/847c783e05/osx_arm64/icu.duckdb_extension.gz" (gzip compressed version) or
     "https://carlopi.github.io/extension-repository/847c783e05/osx_arm64/icu.duckdb_extension" (no encoding)

'LOAD httpfs' might help.
D LOAD httpfs;
D FORCE INSTALL icu; LOAD icu;
D FORCE INSTALL tpch; LOAD tpch;
D FORCE INSTALL tpcds; LOAD tpcds;
D FORCE INSTALL zzz;
Error: IO Error: Failed to download extension "zzz" at either of:
     "https://carlopi.github.io/extension-repository/0fd6fb9198/osx_arm64/zzz.duckdb_extension.gz" (gzip compressed version) or
     "https://carlopi.github.io/extension-repository/0fd6fb9198/osx_arm64/zzz.duckdb_extension" (no encoding)

'LOAD httpfs' might help.
```

All but the `httpfs` extension (that is loaded from the locally build folder) are loaded over https from the GitHub Page hosted at https://carlopi.github.io/extension-repository, that can only accessed over https, backed by the repository with this content: https://github.com/carlopi/extension-repository/tree/main/0fd6fb9198/osx_arm64.

Extension are first looked up in the gzip-compressed version, and if not present in the non-compressed one.
This holds both for when custom_extension_repository is a remote OR local directory.

See for example:
```
carlo@ScroogeMcDuck duckdb2 % ./build/release/duckdb -unsigned
v0.9.3-dev897 0fd6fb9198
Enter ".help" for usage hints.
D SET custom_extension_repository= 'extension-repository';
D FORCE INSTALL tpch; LOAD tpch;
D FORCE INSTALL tpcds; LOAD tpcds;
D FORCE INSTALL icu; LOAD icu;
D FORCE INSTALL zzz;
Error: IO Error: Failed to copy local extension "zzz" at either of:
     "extension-repository/0fd6fb9198/osx_arm64/zzz.duckdb_extension.gz" (gzip compressed version) or
     "extension-repository/0fd6fb9198/osx_arm64/zzz.duckdb_extension" (no encoding)
```
Where extensions works if at least one of the files `name.duckdb_extension.gz` or `name.duckdb_extension` are present at the expected path.

To be done, but not part of this PR:
- [x] Enabling autoloading for httpfs extension (currently there is some missing `context`, might need to rework interface a bit, to be reviewed)
- [ ] Allow also over https to send user-agent string (currently blocked on exposing arbitrary headers on httpfs filesystem
- [ ] Allowing streamlined publishing of extension-template based repositories over httpfs. 

This is still allowing to host repositories on https://, with the only caveat that `httpfs` extension has to be either already present on the system (being installed before hand from the official extension repository).

This also allows to simplify building and using of local extensions, given that once passed `SET custom_extension_repository='build/release'`, (and with the infrastructure discussed in https://github.com/duckdb/duckdb/pull/9800), it would be possible to make autoloading and autoinstall work in a similar way to what users will experience. This allows simplifications of infrastructure.

carlopi committed:
ExtensionIsLocallyInstalled helper function

carlopi committed:
Add restrictions to SET custom_extension_repository = 'https://something.org/'

To be allowed to set CustomExtensionRepository to a remote extension repository requiring https or s3 protocol
httpfs to be available, so for this to succeed httpfs needs to be:
      1. already loaded (potentially built-in)
      2. available to be installed at the local_extension_path

The problem is that allowing to change the custom_extension_repository to something needing httpfs allows users
  to corner themselves out.
Potentially this can be relaxed, but given until serving extensions over https:// was not really working, we can as
  well be strict and relax this later.

carlopi committed:
Allow INSTALL from httpfs-enabled paths

There are, as of now, no tests for this, but for the fact that the rest
of the system keeps working

carlopi committed:
Format fix

carlopi committed:
More format fix

samansmink commented on src/main/extension/extension_install.cpp:
```diff
@@ -244,20 +244,46 @@ void ExtensionHelper::InstallExtensionInternal(DBConfig &config, ClientConfig *c
+					                  "compressed version) or\n     \"%s\" (no encoding)\n%s",
+					                  message_verb, extension_name, file, file, addendum);
+				}
+				endsInGz = false;
```
I think this code is slightly obfuscated, removing endsInGz en rechecking the StringUtil::EndsWith(file, ".gz") in the if statement that now checks endsInGz would make things easier to read.

carlopi replied:
Thanks, improved.

samansmink reviewed (commented):
Looks good!  one veeery minor nitpick

carlopi committed:
Refactor after PR comments

carlopi committed:
Fix: C-cast to reinterpret_cast

carlopi commented:
I think this can be merged, can be polished a bit by allowing autoloading of httpfs, but that's also OK in a separate step since in any case this expand what was previously possible, fixing some corner cases in the process.

While writing the docs I also realised that we are not erroring out nicely on:
```
LOAD 'https://carlopi.github.io/extension-repository/0fd6fb9198/osx_arm64/json.duckdb_extension';
```

carlopi committed:
Expose LookupExtensionForPatter utility function

carlopi committed:
Handle autoloading at the InstallExtension level, removing warnings

carlopi committed:
Merge branch 'main' into httpfsextensions

carlopi commented:
Autoloading of httpfs (or any other extension required by file system), has been handled, even though the used API for doing this might need to be reviewed.

carlopi commented:
There is a funny corner case to be handled, fixing reentrancy.

carlopi committed:
[Try]AutoLoadExtension, add option to avoid install

carlopi committed:
Check for reentrancy in InstallExtension calls

Idea is that it should not be possible to ever go back again in the InstallExtension
codepath.

carlopi committed:
clang-tidy casing change

Mytherin reviewed (commented):
Thanks for the PR! Comments below

carlopi commented:
Solved/superseeded by https://github.com/duckdb/duckdb/pull/11677
